app/src/main/java/com/example/tasky/feature_authentication/presentation/login_screen/LoginViewModel.kt

```diff
@@ -0,0 +1,90 @@
+package com.example.tasky.feature_authentication.presentation.login_screen
+
+import androidx.compose.runtime.State
+import androidx.compose.runtime.mutableStateOf
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.tasky.feature_authentication.domain.use_case.AuthUseCases
+import com.example.tasky.feature_authentication.domain.util.AuthUseCaseResult
+import com.example.tasky.feature_authentication.domain.validation.UserDataValidator
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val useCases: AuthUseCases,
+    private val userDataValidator: UserDataValidator
+) : ViewModel() {
+    private val _state = mutableStateOf(LoginState())
+    val state: State<LoginState> = _state
+
+    private val resultChannel = Channel<AuthUseCaseResult>()
+    val authResult = resultChannel.receiveAsFlow()
+
+    private val _isLoading = MutableStateFlow(true)
+    val isLoading = _isLoading.asStateFlow()
+    init {
+        authenticate()
+        _isLoading.value = false
+    }
+
+    fun onEvent(event: LoginEvent) {
+        when(event) {
+            is LoginEvent.ChangePasswordVisibility -> {
+                _state.value = state.value.copy(
+                    isPasswordVisible = !state.value.isPasswordVisible
+                )
+            }
+            is LoginEvent.EmailChanged -> {
+                _state.value = state.value.copy(
+                    email = event.email
+                )
+                validateData(state.value.email)
+            }
+            is LoginEvent.PasswordChanged -> {
+                _state.value = state.value.copy(
+                    password = event.password
+                )
+            }
+            is LoginEvent.SignIn -> {
+                signIn()
+            }
+        }
+    }
+
+    private fun validateData(email: String) {
+        val result = userDataValidator.validateEmail(email)
+
+        if(result.isValid) {
+            _state.value = state.value.copy(
+                isEmailValid = true
+            )
+        } else {
+            _state.value = state.value.copy(
+                isEmailValid = false
+            )
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue (and in your case it isn't), but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.
Using `_state.value = _state.value.copy(...)` is fine if you always call this from a single thread (which you do here). But in ViewModels, you're very likely to not do that as soon as you'd change the state inside a `viewModelScope`.